csharp/extractor/Semmle.Extraction.CSharp/Entities/Types/Type.cs

```diff
@@ -347,6 +355,211 @@ public override bool Equals(object? obj)
         }
 
         public override int GetHashCode() => SymbolEqualityComparer.Default.GetHashCode(Symbol);
+
+        /// <summary>
+        /// Class to detect recursive generic inheritance hierarchies.
+        ///
+        /// Details can be found in https://www.ecma-international.org/wp-content/uploads/ECMA-335_6th_edition_june_2012.pdf Chapter II.9.2 Generics and recursive inheritance graphs
+        /// The dotnet runtime already implements this check as a runtime validation: https://github.com/dotnet/runtime/blob/e48e88d0fe9c2e494c0e6fd0c7c1fb54e7ddbdb1/src/coreclr/vm/generics.cpp#L748
+        /// </summary>
+        public class GenericsRecursionGraph
+        {
+            private static readonly ConcurrentDictionary<INamedTypeSymbol, bool> resultCache = new(SymbolEqualityComparer.Default);
+
+            /// <summary>
+            /// Checks whether the given type has a recursive generic inheritance hierarchy. The result is cached.
+            /// </summary>
+            public static bool HasExpandingCycle(ITypeSymbol start)
+            {
+                if (start.OriginalDefinition is not INamedTypeSymbol namedTypeDefinition ||
+                    !namedTypeDefinition.IsGenericType)
+                {
+                    return false;
+                }
+
+                if (resultCache.TryGetValue(namedTypeDefinition, out var result))
+                {
+                    return result;
+                }
+
+                result = new GenericsRecursionGraph(namedTypeDefinition).HasExpandingCycle();
+                resultCache.TryAdd(namedTypeDefinition, result);
+
+                return result;
+            }
+
+            private readonly INamedTypeSymbol startSymbol;
+            private readonly HashSet<INamedTypeSymbol> instantiationClosure = new(SymbolEqualityComparer.Default);
+            private readonly Dictionary<ITypeParameterSymbol, List<(ITypeParameterSymbol To, bool IsExpanding)>> edges = new(SymbolEqualityComparer.Default);
+
+            private GenericsRecursionGraph(INamedTypeSymbol startSymbol)
+            {
+                this.startSymbol = startSymbol;
+
+                ComputeInstantiationClosure();
+                ComputeGraphEdges();
+            }
+
+            private void ComputeGraphEdges()
+            {
+                foreach (var reference in instantiationClosure)
+                {
+                    var definition = reference.OriginalDefinition;
+                    if (SymbolEqualityComparer.Default.Equals(reference, definition))
```

Comment:
Use `IsReallyUnbound`?

Reply:
Thanks. I should check what happens with unbound types, shouldn't I?

So basically I should check the processing of `GenB<>` in the below. Where else can unbound generic types show up?

```C#
public class GenA<U> { };
public class GenB<T> : GenA<GenB<GenB<T>>> { };

class AAttribute : System.Attribute
{
    public AAttribute(System.Type t){}
}

public class Class
{
    [A(typeof(GenB<>))]
    public static int Main()
    {
        //GenB<string> a = new GenB<string>();

        return 0;
    }
}
```

Reply:
@hvitved I checked the above, and I don't think we can face an unbound generic type inside the `GenericsRecursionGraph`, because the constructor is invoked with the original definition of a named type symbol, and the original definition of an unbound generic type is not unbound.